### PR TITLE
Package camomile.2.0.0

### PR DESCRIPTION
**Pull request by toots**
### `camomile.2.0.0`
A Unicode library
Camomile is a Unicode library for OCaml. Camomile provides Unicode character
type, UTF-8, UTF-16, UTF-32 strings, conversion to/from about 200 encodings,
collation and locale-sensitive case mappings, and more. The library is currently
designed for Unicode Standard 3.2.

---
* Homepage: https://github.com/savonet/Camomile
* Source repo: git+https://github.com/savonet/Camomile.git
* Bug tracker: https://github.com/savonet/Camomile/issues

---
:camel: Pull-request generated by opam-publish v2.2.0

### CODE REVIEW

**Comment by kit-ty-kate** [comment]
most revdeps seem to fail with a variant of
```
#=== ERROR while compiling soupault.4.1.0 =====================================#
# context              2.2.0~alpha~dev | linux/x86_64 | ocaml-base-compiler.4.14.1 | file:///home/opam/opam-repository
# path                 ~/.opam/4.14/.opam-switch/build/soupault.4.1.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p soupault -j 47
# exit-code            1
# env-file             ~/.opam/log/soupault-7-1cc480.env
# output-file          ~/.opam/log/soupault-7-1cc480.out
### output ###
# (cd _build/default && /home/opam/.opam/4.14/bin/ocamlc.opt -w -40 -g -bin-annot -I src/.soupault.eobjs/byte -I /home/opam/.opam/4.14/lib/base64 -I /home/opam/.opam/4.14/lib/bigarray-compat -I /home/opam/.opam/4.14/lib/camlp-streams -I /home/opam/.opam/4.14/lib/camomile -I /home/opam/.opam/4.14/lib/camomile/lib -I /home/opam/.opam/4.14/lib/containers -I /home/opam/.opam/4.14/lib/containers/monomorphic -I /home/opam/.opam/4.14/lib/cstruct -I /home/opam/.opam/4.14/lib/ctypes -I /home/opam/.opam/4.14/lib/dune-private-libs/dune-section -I /home/opam/.opam/4.14/lib/dune-site -I /home/opam/.opam/4.14/lib/dune-site/private -I /home/opam/.opam/4.14/lib/either -I /home/opam/.opam/4.14/lib/ezjsonm -I /home/opam/.opam/4.14/lib/fileutils -I /home/opam/.opam/4.14/lib/fmt -I /home/opam/.opam/4.14/lib/hex -I /home/opam/.opam/4.14/lib/integers -I /home/opam/.opam/4.14/lib/jingoo -I /home/opam/.opam/4.14/lib/jsonm -I /home/opam/.opam/4.14/lib/lambdasoup -I /home/opam/.opam/4.14/lib/logs -I /home/opam/.opam/4.14/lib/lua-ml -I /home/opam/.opam/4.14/lib/markup -I /home/opam/.opam/4.14/lib/menhirLib -I /home/opam/.opam/4.14/lib/odate -I /home/opam/.opam/4.14/lib/otoml -I /home/opam/.opam/4.14/lib/ppx_deriving/runtime -I /home/opam/.opam/4.14/lib/re -I /home/opam/.opam/4.14/lib/result -I /home/opam/.opam/4.14/lib/seq -I /home/opam/.opam/4.14/lib/sexplib0 -I /home/opam/.opam/4.14/lib/spelll -I /home/opam/.opam/4.14/lib/stdlib-shims -I /home/opam/.opam/4.14/lib/tsort -I /home/opam/.opam/4.14/lib/uucp -I /home/opam/.opam/4.14/lib/uutf -I /home/opam/.opam/4.14/lib/yaml -I /home/opam/.opam/4.14/lib/yaml/bindings -I /home/opam/.opam/4.14/lib/yaml/bindings/types -I /home/opam/.opam/4.14/lib/yaml/c -I /home/opam/.opam/4.14/lib/yaml/ffi -I /home/opam/.opam/4.14/lib/yaml/types -no-alias-deps -open Dune__exe -o src/.soupault.eobjs/byte/dune__exe__Text.cmo -c -impl src/text.ml)
# File "src/text.ml", line 1, characters 5-27:
# 1 | open CamomileLibraryDefault
#          ^^^^^^^^^^^^^^^^^^^^^^
# Error: Unbound module CamomileLibraryDefault
```

**Comment by toots** [comment]
> most revdeps seem to fail with a variant of
> 
> ```
> #=== ERROR while compiling soupault.4.1.0 =====================================#
> # context              2.2.0~alpha~dev | linux/x86_64 | ocaml-base-compiler.4.14.1 | file:///home/opam/opam-repository
> # path                 ~/.opam/4.14/.opam-switch/build/soupault.4.1.0
> # command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p soupault -j 47
> # exit-code            1
> # env-file             ~/.opam/log/soupault-7-1cc480.env
> # output-file          ~/.opam/log/soupault-7-1cc480.out
> ### output ###
> # (cd _build/default && /home/opam/.opam/4.14/bin/ocamlc.opt -w -40 -g -bin-annot -I src/.soupault.eobjs/byte -I /home/opam/.opam/4.14/lib/base64 -I /home/opam/.opam/4.14/lib/bigarray-compat -I /home/opam/.opam/4.14/lib/camlp-streams -I /home/opam/.opam/4.14/lib/camomile -I /home/opam/.opam/4.14/lib/camomile/lib -I /home/opam/.opam/4.14/lib/containers -I /home/opam/.opam/4.14/lib/containers/monomorphic -I /home/opam/.opam/4.14/lib/cstruct -I /home/opam/.opam/4.14/lib/ctypes -I /home/opam/.opam/4.14/lib/dune-private-libs/dune-section -I /home/opam/.opam/4.14/lib/dune-site -I /home/opam/.opam/4.14/lib/dune-site/private -I /home/opam/.opam/4.14/lib/either -I /home/opam/.opam/4.14/lib/ezjsonm -I /home/opam/.opam/4.14/lib/fileutils -I /home/opam/.opam/4.14/lib/fmt -I /home/opam/.opam/4.14/lib/hex -I /home/opam/.opam/4.14/lib/integers -I /home/opam/.opam/4.14/lib/jingoo -I /home/opam/.opam/4.14/lib/jsonm -I /home/opam/.opam/4.14/lib/lambdasoup -I /home/opam/.opam/4.14/lib/logs -I /home/opam/.opam/4.14/lib/lua-ml -I /home/opam/.opam/4.14/lib/markup -I /home/opam/.opam/4.14/lib/menhirLib -I /home/opam/.opam/4.14/lib/odate -I /home/opam/.opam/4.14/lib/otoml -I /home/opam/.opam/4.14/lib/ppx_deriving/runtime -I /home/opam/.opam/4.14/lib/re -I /home/opam/.opam/4.14/lib/result -I /home/opam/.opam/4.14/lib/seq -I /home/opam/.opam/4.14/lib/sexplib0 -I /home/opam/.opam/4.14/lib/spelll -I /home/opam/.opam/4.14/lib/stdlib-shims -I /home/opam/.opam/4.14/lib/tsort -I /home/opam/.opam/4.14/lib/uucp -I /home/opam/.opam/4.14/lib/uutf -I /home/opam/.opam/4.14/lib/yaml -I /home/opam/.opam/4.14/lib/yaml/bindings -I /home/opam/.opam/4.14/lib/yaml/bindings/types -I /home/opam/.opam/4.14/lib/yaml/c -I /home/opam/.opam/4.14/lib/yaml/ffi -I /home/opam/.opam/4.14/lib/yaml/types -no-alias-deps -open Dune__exe -o src/.soupault.eobjs/byte/dune__exe__Text.cmo -c -impl src/text.ml)
> # File "src/text.ml", line 1, characters 5-27:
> # 1 | open CamomileLibraryDefault
> #          ^^^^^^^^^^^^^^^^^^^^^^
> # Error: Unbound module CamomileLibraryDefault
> ```

Yes, the API has been cleaned up and all revdeps need to conflict with the new version. I'll add it.

**Comment by toots** [comment]
@kit-ty-kate this should be looking good now. I'm not sure why this one test is failing on OCaml 5 yet. It's base on random-generated input and I think the error is that the random range has changed between version so it seems safe to ignore for now.

This release is strictly cleanup, I don't have to change the implementation yet.

**Comment by kit-ty-kate** [comment]
In that case those are genuine errors on OCaml 5.0 and thus the tests should be disabled on OCaml 5.0 until camomile uses https://github.com/ocaml/stdlib-random or equivalent.

File: packages/camomile/camomile.2.0.0/opam [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  [
+    "dune"
+    "build"
+    "-p"
+    name
+    "-j"
+    jobs
+    "--promote-install-files=false"
+    "@install"
+    "@runtest" {with-test}
```
Comment (kit-ty-kate) [inline packages/camomile/camomile.2.0.0/opam:33]:
```suggestion
    "@runtest" {with-test & ocaml:version < "5.0"}
```

**Comment by toots** [comment]
> It looks pretty good apart this segfault on ppc64:
> 
> ```
> #=== ERROR while compiling camomile.2.0.0 =====================================#
> # context              2.2.0~alpha~dev | linux/ppc64 | ocaml-base-compiler.4.14.1 | pinned(https://github.com/savonet/Camomile/archive/refs/tags/v2.0.0.tar.gz)
> # path                 ~/.opam/4.14/.opam-switch/build/camomile.2.0.0
> # command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p camomile -j 175 --promote-install-files=false @install
> # exit-code            1
> # env-file             ~/.opam/log/camomile-6-ec6679.env
> # output-file          ~/.opam/log/camomile-6-ec6679.out
> ### output ###
> # File "src/locales/dune.inc", line 1967, characters 0-142:
> # 1967 | (rule
> # 1968 |  (targets zh__PINYIN.mar)
> # 1969 |  (deps    (:x zh__PINYIN.txt) (alias database))
> # 1970 |  (action  (run ../tools/camomilelocaledef.exe --file %{x} .)))
> # (cd _build/default/src/locales && ../tools/camomilelocaledef.exe --file zh__PINYIN.txt .)
> # Command got signal SEGV.
> # (cd _build/default/src/locales && ../tools/camomilelocaledef.exe --file ja.txt .)
> # Warning : strength option is not supported
> ```
> 
> I didn't test locally but I'm guessing this has to do with `zh__PINYIN.txt` being the largest files in locale and some sort of stack overflow that isn't caught properly on ppc64 (expected). Do you want to have a look at it first or do you want to ignore that?

I think we can let it go for now. I want to revisit the database layer in the next phase to see if we could find a way to have all these values embedded. As far as I can see they are marshaled values so the only benefit of the files is to not have the always loaded in memory.

**Comment by kit-ty-kate** [comment]
Thanks

**Pull request merged**